# NYC Citywide Payroll

In this case study, we are going to dig into the New York citywide payroll data and illustrate the highest paying job categories using a sunburst diagram. Let's start with the data.

## Getting the data

Our data source is [NYC open data portal](https://data.cityofnewyork.us/City-Government/Citywide-Payroll-Data-Fiscal-Year-/k397-673e). It contains the salary/wage data of every employee hired by the city government since around 2015.

In [1]:
import altair as alt
import pandas as pd

In [2]:
data = pd.read_csv("https://data.cityofnewyork.us/api/views/k397-673e/rows.csv?accessType=DOWNLOAD")
data

/var/folders/xj/gx6zl13s2gb92t9fvm42xgdh0000gn/T/ipykernel_9372/4286039976.py:1: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("https://data.cityofnewyork.us/api/views/k397-673e/rows.csv?accessType=DOWNLOAD")


Fiscal Year  Payroll Number                     Agency Name  \
0               2020            17.0  OFFICE OF EMERGENCY MANAGEMENT   
1               2020            17.0  OFFICE OF EMERGENCY MANAGEMENT   
2               2020            17.0  OFFICE OF EMERGENCY MANAGEMENT   
3               2020            17.0  OFFICE OF EMERGENCY MANAGEMENT   
4               2020            17.0  OFFICE OF EMERGENCY MANAGEMENT   
...              ...             ...                             ...   
5109770         2022             2.0             OFFICE OF THE MAYOR   
5109771         2022             2.0             OFFICE OF THE MAYOR   
5109772         2022             2.0             OFFICE OF THE MAYOR   
5109773         2022             2.0             OFFICE OF THE MAYOR   
5109774         2022             2.0             OFFICE OF THE MAYOR   

         Last Name First Name Mid Init Agency Start Date  \
0          BEREZIN    MIKHAIL      NaN        08/10/2015   
1           GEAGER   VERONICA        M        09/12/2016   
2           RAMANI   SHRADDHA      NaN        02/22/2016   
3            ROTTA   JONATHAN        D        09/16/2013   
4        WILSON II     ROBERT        P        04/30/2018   
...            ...        ...      ...               ...   
5109770      ZHANG        JIN      NaN        12/09/2019   
5109771       ZHAO    XIAOMIN      NaN        09/04/2018   
5109772      ZHENG      SANDY      NaN        06/08/2022   
5109773       ZHOU      DIANE        J        11/13/2017   
5109774       ZITA      JULIA        L        08/23/2021   

        Work Location Borough               Title Description  \
0                    BROOKLYN  EMERGENCY PREPAREDNESS MANAGER   
1                    BROOKLYN  EMERGENCY PREPAREDNESS MANAGER   
2                    BROOKLYN  EMERGENCY PREPAREDNESS MANAGER   
3                    BROOKLYN  EMERGENCY PREPAREDNESS MANAGER   
4                    BROOKLYN  EMERGENCY PREPAREDNESS MANAGER   
...                       ...                             ...   
5109770             MANHATTAN          Senior Project Planner   
5109771             MANHATTAN               SPECIAL ASSISTANT   
5109772             MANHATTAN           SUMMER COLLEGE INTERN   
5109773             MANHATTAN        MAYORAL OFFICE ASSISTANT   
5109774             MANHATTAN               SPECIAL ASSISTANT   

        Leave Status as of June 30  Base Salary  Pay Basis  Regular Hours  \
0                           ACTIVE     86005.00  per Annum        1820.00   
1                           ACTIVE     86005.00  per Annum        1820.00   
2                           ACTIVE     86005.00  per Annum        1820.00   
3                           ACTIVE     86005.00  per Annum        1820.00   
4                           ACTIVE     86005.00  per Annum        1820.00   
...                            ...          ...        ...            ...   
5109770                     CEASED    110000.00  per Annum          49.00   
5109771                     CEASED     71974.00  per Annum         980.00   
5109772                     ACTIVE        15.75   per Hour          22.25   
5109773                     ACTIVE     50000.00  per Annum        1820.00   
5109774                     CEASED     77000.00  per Annum         672.00   

         Regular Gross Paid  OT Hours  Total OT Paid  Total Other Pay  
0                  84698.21       0.0            0.0             0.00  
1                  84698.21       0.0            0.0             0.00  
2                  84698.21       0.0            0.0             0.00  
3                  84698.21       0.0            0.0             0.00  
4                  84698.21       0.0            0.0             0.00  
...                     ...       ...            ...              ...  
5109770             2953.16       0.0            0.0          1883.56  
5109771            38649.10       0.0            0.0             0.00  
5109772              350.44       0.0            0.0             0.00

This dataset actually contains quite a lot of information such as the employee name, job title, agency, borough, base salary, etc.  For this case study, we are mainly interested in the `Base Salary` column. Note that `Base Salary` actually may encode two types of information: annual salary and hourly wage. The type is specified in the `Pay Basis` column.

Just as a sanity check, let us look up the salary of New York City mayor. :)

In [3]:
data[data['Title Description']=="MAYOR"]

Fiscal Year  Payroll Number          Agency Name  Last Name  \
148859          2020             2.0  OFFICE OF THE MAYOR  DE BLASIO   
549134          2021             2.0  OFFICE OF THE MAYOR  DE BLASIO   
1712979         2016             NaN  OFFICE OF THE MAYOR  DE BLASIO   
2283359         2015             NaN  OFFICE OF THE MAYOR  DE BLASIO   
2288546         2014             2.0  OFFICE OF THE MAYOR  BLOOMBERG   
2288959         2014             2.0  OFFICE OF THE MAYOR  DE BLASIO   
3288779         2017             NaN  OFFICE OF THE MAYOR  DE BLASIO   
3302127         2018             2.0  OFFICE OF THE MAYOR  DE BLASIO   
3907283         2019             2.0  OFFICE OF THE MAYOR  DE BLASIO   
5109041         2022             2.0  OFFICE OF THE MAYOR      ADAMS   
5109221         2022             2.0  OFFICE OF THE MAYOR  DE BLASIO   

        First Name Mid Init Agency Start Date Work Location Borough  \
148859        BILL      NaN        01/01/2014             MANHATTAN   
549134        BILL      NaN        01/01/2014             MANHATTAN   
1712979       BILL      NaN        01/01/2014             MANHATTAN   
2283359       BILL      NaN        01/01/2014             MANHATTAN   
2288546    MICHAEL        R        01/01/2002                   NaN   
2288959       BILL      NaN        01/01/2014                   NaN   
3288779       BILL      NaN        01/01/2014             MANHATTAN   
3302127       BILL      NaN        01/01/2014             MANHATTAN   
3907283       BILL      NaN        01/01/2014             MANHATTAN   
5109041       ERIC        L        01/01/2022             MANHATTAN   
5109221       BILL      NaN        01/01/2014             MANHATTAN   

        Title Description Leave Status as of June 30  Base Salary  Pay Basis  \
148859              MAYOR                     ACTIVE     258750.0  per Annum   
549134              MAYOR                     ACTIVE     258750.0  per Annum   
1712979             MAYOR                     ACTIVE     225000.0  per Annum   
2283359             MAYOR                     ACTIVE     225000.0  per Annum   
2288546             MAYOR                     CEASED          1.0  per Annum   
2288959             MAYOR                     ACTIVE     225000.0  per Annum   
3288779             MAYOR                     ACTIVE     225000.0  per Annum   
3302127             MAYOR                     ACTIVE     258750.0  per Annum   
3907283             MAYOR                     ACTIVE     258750.0  per Annum   
5109041             MAYOR                     ACTIVE     258750.0  per Annum   
5109221             MAYOR                     CEASED     258750.0  per Annum   

         Regular Hours  Regular Gross Paid  OT Hours  Total OT Paid  \
148859          1820.0           257351.53       0.0            0.0   
549134          1820.0           258021.79       0.0            0.0   
1712979         1830.0           223799.19       0.0            0.0   
2283359         1825.0           224383.64       0.0            0.0   
2288546            1.0                1.00       0.0            0.0   
2288959          906.0           101835.64       0.0            0.0   
3288779         1825.0           232981.78       0.0            0.0   
3302127         1790.0           240565.14       0.0            0.0   
3907283         1825.0           258041.16       0.0            0.0   
5109041          842.0           120088.35       0.0            0.0   
5109221          980.0           138945.24       0.0            0.0   

         Total Other Pay  
148859            1500.0  
549134           -4956.9  
1712979              0.0  
2283359              0.0  
2288546              0.0  
2288959              0.0  
3288779              0.0  
3302127           3000.0  
3907283            500.0  
5109041              0.0  
5109221           5456.9

As you can see from the result, the dataset contains pay information of three different mayors.  Michael Bloomberg (2014), Bill DeBlasio (from 2014 to 2022) and Eric Adams (2022).  The 2022 base salary of Eric Adams is \$258,750. This is consistent with the number stated on [Wikipedia](https://en.wikipedia.org/wiki/Mayor_of_New_York_City). Another interesting fact is that Michael Bloomberg's base salary in 2014 is just \$1, which is not a typo!  The reason is also explained in the [same Wikipedia article](https://en.wikipedia.org/wiki/Mayor_of_New_York_City).

## Data wrangling

While the dataset is vast and quite interesting, let us focus on the 2022 annual salary data in this case study.

In [4]:
active_annual = data[(data["Pay Basis"] == "per Annum") & \
                     (data["Leave Status as of June 30"] == "ACTIVE") & \
                     (data["Fiscal Year"] == 2022)]
active_annual

Fiscal Year  Payroll Number             Agency Name   Last Name  \
2530            2022           742.0  DEPT OF ED PEDAGOGICAL    APPELMAN   
3162            2022           742.0  DEPT OF ED PEDAGOGICAL    ARBOLINO   
3611            2022           742.0  DEPT OF ED PEDAGOGICAL     ARCIERO   
4296            2022           742.0  DEPT OF ED PEDAGOGICAL      ARCILA   
4623            2022           742.0  DEPT OF ED PEDAGOGICAL  ARCINIEGAS   
...              ...             ...                     ...         ...   
5109765         2022             2.0     OFFICE OF THE MAYOR          YE   
5109766         2022             2.0     OFFICE OF THE MAYOR       YOUNG   
5109767         2022             2.0     OFFICE OF THE MAYOR       YOUNG   
5109769         2022             2.0     OFFICE OF THE MAYOR  ZAFIRIADIS   
5109773         2022             2.0     OFFICE OF THE MAYOR        ZHOU   

         First Name Mid Init Agency Start Date Work Location Borough  \
2530         BROOKE        A        10/20/2015             MANHATTAN   
3162         ARIANA        M        09/05/2000             MANHATTAN   
3611         DENISE        A        08/30/2007             MANHATTAN   
4296          MARIA        I        02/02/2022             MANHATTAN   
4623         KARINA        B        09/06/2005             MANHATTAN   
...             ...      ...               ...                   ...   
5109765       JENNY        J        03/17/2021             MANHATTAN   
5109766     MAXWELL        C        01/08/2022             MANHATTAN   
5109767      SKYLAR        V        10/25/2021             MANHATTAN   
5109769  CHRISTIANA      NaN        12/05/2016             MANHATTAN   
5109773       DIANE        J        11/13/2017             MANHATTAN   

                Title Description Leave Status as of June 30  Base Salary  \
2530                      TEACHER                     ACTIVE      97469.0   
3162                    PRINCIPAL                     ACTIVE     169456.0   
3611                      TEACHER                     ACTIVE     100351.0   
4296                      TEACHER                     ACTIVE      69866.0   
4623                      TEACHER                     ACTIVE     109852.0   
...                           ...                        ...          ...   
5109765   RESEARCH PROJECTS COORD                     ACTIVE      97630.0   
5109766             PRESS OFFICER                     ACTIVE     211150.0   
5109767         SPECIAL ASSISTANT                     ACTIVE      60000.0   
5109769         SPECIAL ASSISTANT                     ACTIVE     115000.0   
5109773  MAYORAL OFFICE ASSISTANT                     ACTIVE      50000.0   

         Pay Basis  Regular Hours  Regular Gross Paid  OT Hours  \
2530     per Annum            0.0            90693.25       0.0   
3162     per Annum            0.0           168222.06       0.0   
3611     per Annum            0.0           100350.96       0.0   
4296     per Annum            0.0            26274.50       0.0   
4623     per Annum            0.0           104620.70       0.0   
...            ...            ...                 ...       ...   
5109765  per Annum         1820.0            89882.23       0.0   
5109766  per Annum          842.0            93947.23       0.0   
5109767  per Annum         1190.0            39123.29       0.0   
5109769  per Annum         1820.0           110336.16       0.0   
5109773  per Annum         1820.0            49863.06       0.0   

         Total OT Paid  Total Other Pay  
2530               0.0          1486.01  
3162               0.0         16947.18  
3611               0.0          4851.73  
4296               0.0          2801.02  
4623               0.0         18903.82  
...                ...              ...  
5109765            0.0             0.00  
5109766            0.0             0.00  
5109767            0.0             0.00  
5109769            0.0          1996.20  
5109773            0.0             0.00

Here, we created a sub-dataset of 2022 payroll data for all active employees receiving per annum pay.  This sub-dataset still contains quite a lot of information. Let us further narrow down our focus on the 10 largest agencies in terms of headcount.

In [5]:
largest_agency = active_annual.groupby(['Agency Name'], dropna=True)['Base Salary']\
    .count().sort_values(ascending=False)[:10]
largest_agency

Agency Name
DEPT OF ED PEDAGOGICAL           93986
POLICE DEPARTMENT                47424
DEPT OF ED PARA PROFESSIONALS    23366
FIRE DEPARTMENT                  16474
DEPARTMENT OF EDUCATION ADMIN    11388
HRA/DEPT OF SOCIAL SERVICES      10542
DEPARTMENT OF SANITATION          8757
NYC HOUSING AUTHORITY             8644
DEPARTMENT OF CORRECTION          7914
ADMIN FOR CHILDREN'S SVCS         6188
Name: Base Salary, dtype: int64

Not surprisingly, education, police, fire and housing departments are among the largest agencies.  Furthermore, the largest agency is nearly twice as large as the second largest agency. 

For this case study, we would like to visualize the salary data at two different hierarchical levels: by agency and by job title. To do this, we need to create an aggregate dataset for each hierarchical level.

In [6]:
largest_agency_salary = active_annual[active_annual['Agency Name'].isin(largest_agency.index)]
salary_by_agency = largest_agency_salary.groupby(
    ["Agency Name"], dropna=True, as_index=False)['Base Salary'].agg(['count', 'mean'])
salary_by_agency = salary_by_agency.reset_index()
salary_by_agency.columns = ['agency', 'count', 'mean']
salary_by_agency

agency  count           mean
0      ADMIN FOR CHILDREN'S SVCS   6188   71295.894958
1       DEPARTMENT OF CORRECTION   7914   87618.959692
2  DEPARTMENT OF EDUCATION ADMIN  11388   75279.817966
3       DEPARTMENT OF SANITATION   8757   79347.844211
4  DEPT OF ED PARA PROFESSIONALS  23366   40259.478687
5         DEPT OF ED PEDAGOGICAL  93986  100739.758443
6                FIRE DEPARTMENT  16474   82854.382603
7    HRA/DEPT OF SOCIAL SERVICES  10542   61579.625593
8          NYC HOUSING AUTHORITY   8644   58820.807612
9              POLICE DEPARTMENT  47424   76661.791983

This `salary_by_agency` dataset just contains three columns: the agency name, its number of employees and their average salary. We can create the `salary_by_title` dataset in a similar manner.

In [7]:
salary_by_title = largest_agency_salary.groupby(
    ["Agency Name", "Title Description"], dropna=True, as_index=False)['Base Salary'].agg(['count', 'mean'])
salary_by_title = salary_by_title.reset_index()
salary_by_title.columns = ["agency", "title", "count", "mean"]
salary_by_title

agency                                        title  \
0     ADMIN FOR CHILDREN'S SVCS                             *ATTORNEY AT LAW   
1     ADMIN FOR CHILDREN'S SVCS            *CERTIFIED DATABASE ADMINISTRATOR   
2     ADMIN FOR CHILDREN'S SVCS  *CERTIFIED LOCAL AREA NETWORK ADMINISTRATOR   
3     ADMIN FOR CHILDREN'S SVCS                                        *COOK   
4     ADMIN FOR CHILDREN'S SVCS                                 *SENIOR COOK   
...                         ...                                          ...   
1243          POLICE DEPARTMENT                  SUPERVISOR OF STOCK WORKERS   
1244          POLICE DEPARTMENT                 TELECOMMUNICATIONS ASSOCIATE   
1245          POLICE DEPARTMENT                 TELEPHONE SERVICE TECHNICIAN   
1246          POLICE DEPARTMENT             TESTS AND MEASUREMENT SPECIALIST   
1247          POLICE DEPARTMENT                    TRAFFIC ENFORCEMENT AGENT   

      count           mean  
0         6   96916.166667  
1         2  115599.000000  
2         1  122075.000000  
3         9   40885.555556  
4         3   42937.666667  
...     ...            ...  
1243      8   54790.250000  
1244     12   86919.166667  
1245      3   75048.333333  
1246      1  105138.000000  
1247   2455   44485.668024  

[1248 rows x 4 columns]

## Sunburst Diagram

One way to think about a sunburst diagram is that it is basically a set of nested donut diagrams. The donut diagrams visualize the same data at different hierarchies.  Since our data contains two hierarchies, we only need to create two donut diagrams.

In [8]:
inner_ring = alt.Chart(salary_by_agency).mark_arc(innerRadius=50, outerRadius=100)\
.encode(theta="count:Q", 
        color="agency:N", 
        tooltip=['agency', 'mean'],
        order="agency")
inner_ring

alt.Chart(...)

We have created the inner donut ring in the above code. Note that we explicitly specify `order="agency"` to control the ordering of the arcs. This is important because we would like the arcs of all donut charts to be consistently ordered. We also added a `tooltip` encoding so that agency and average salary data will be displayed when viewers hover their mouse over each arc.  This is the most basic form of interactive data visualization, which is the topic of another module.

Now let's add an outer ring.

In [9]:
outer_ring = alt.Chart(salary_by_title).mark_arc(innerRadius=100, outerRadius=150, stroke="white")\
.encode(theta="count:Q", 
        color="agency:N", 
        order="agency",
        tooltip=['agency', 'title', 'mean'],
        opacity=alt.condition("datum.mean < 50000 | datum.count < 1000", alt.value(0.25), alt.value(0.75)))
inner_ring + outer_ring

alt.LayerChart(...)

In the outer ring creation, we need to ensure that its inner and outer radii are properly set so it can be composited with the inner ring without overlaps. Furthermore, we also added an opacity encoding based on the headcount and average salary corresponding to a job title. An arc is more transparent if less than 1000 people share this job title or the average salary for this job title is below \$50,000. Otherwise, the arc corresponding to this job title will be more opaque.  Once again, we use tooltip encoding to display relevant information triggered by mouse hovering.

As shown in the visualization, teacher, teacher special education and police officer are job titles with the most headcount that make more than $50,000 annually.

Because our data is intrinsically hierarchical, we can always keep adding more outer rings to our sunburst diagram to visualize more refined hierarchical information. For example, we can visualize the job titles that make more than \$100,000 annually in the third ring.

In [10]:
outer_ring_2 = alt.Chart(salary_by_title).mark_arc(innerRadius=150, outerRadius=200, stroke="white")\
.encode(theta="count:Q", 
        color="agency:N", 
        detail="title:N", 
        order="agency",
        tooltip=["agency", 'title', "mean"],
        opacity=alt.condition("datum.mean < 100000 | datum.count < 1000", alt.value(0), alt.value(0.5)))
inner_ring + outer_ring + outer_ring_2

alt.LayerChart(...)

## Summary

In summary, we have created a sunburst visualization for the NYC citywide payroll dataset. The payroll dataset is implicitly hierarchical. We can create a sub-dataset by aggregation at different hierarchical levels. On the visualization side, the sunburst diagram is basically a set of nested donut charts. Each donut chart visualizes the same data at different hierarchical levels, which can be easily created using Altair. Lastly, we leverage tooltip encoding to provide basic user-intersection to show additional information such as headcount and average salary.